pkg/orchestrator/scheduler/batch_job.go

```diff
@@ -0,0 +1,202 @@
+	err = b.failUnhealthyExecs(ctx, nonTerminalExecutions, nodeInfos, plan)
+	if err != nil {
+		return err
+	}
+
+	job, err := b.jobStore.GetJob(ctx, evaluation.JobID)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve job %s: %w", evaluation.JobID, err)
+	}
+
+	// create new executions if needed
+	activeExecutionsCount := len(nonTerminalExecutions) - plan.StoppedExecutionsCount()
+	desiredExecutionCount := job.Spec.Deal.Concurrency - jobState.CompletedCount() - activeExecutionsCount
+	if desiredExecutionCount > 0 {
+		selectedNodes, err := b.selectNodes(ctx, &job, desiredExecutionCount)
+		if err != nil {
+			return err
+		}
+		for i := 0; i < desiredExecutionCount; i++ {
+			executionID := model.ExecutionID{
+				JobID:       job.Metadata.ID,
+				NodeID:      selectedNodes[i].PeerInfo.ID.String(),
+				ExecutionID: "e-" + uuid.NewString(),
+			}
+			execution := &model.ExecutionState{
+				JobID:            executionID.JobID,
+				NodeID:           executionID.NodeID,
+				ComputeReference: executionID.ExecutionID,
+				State:            model.ExecutionStateAskForBid,
+			}
+			plan.AppendExecution(execution)
+		}
+	}
+	// stop executions if we over-subscribed and exceeded the desired number of executions
+	if desiredExecutionCount < 0 {
+		candidateExecutions := lo.Filter(nonTerminalExecutions, func(execution *model.ExecutionState, _ int) bool {
+			return !plan.IsExecutionStopped(execution)
+		})
+		// TODO: keep track of execution ranks and kill the worst ones
+		// Using version as indicator of which execution has made more progress
+		sort.Slice(candidateExecutions, func(i, j int) bool {
+			return candidateExecutions[i].Version < candidateExecutions[j].Version
+		})
+		for i := 0; i < math.Abs(desiredExecutionCount); i++ {
+			plan.AppendStoppedExecution(candidateExecutions[i], model.ExecutionStateCanceled, execNotNeeded)
+		}
+	}
+
+	return b.planner.Process(ctx, plan)
+}
+
+// failUnhealthyExecs marks executions that are running on nodes that are not healthy as failed
+func (b *BatchJobScheduler) failUnhealthyExecs(
+	ctx context.Context, existingExecutions []*model.ExecutionState, nodeInfos map[string]*model.NodeInfo, plan *models.Plan) error {
+	for _, execution := range existingExecutions {
+		if _, ok := nodeInfos[execution.NodeID]; !ok {
+			plan.AppendStoppedExecution(execution, model.ExecutionStateFailed, execLost)
+		}
+	}
+	return nil
+}
+
+// existingNodeInfos returns a map of nodeID to NodeInfo for all the nodes that have executions for this job
+func (b *BatchJobScheduler) existingNodeInfos(ctx context.Context, existingExecutions []*model.ExecutionState) (map[string]*model.NodeInfo, error) {
+	out := make(map[string]*model.NodeInfo)
+	checked := make(map[string]struct{})
+
+	// TODO: implement a better way to retrieve node info instead of listing all nodes
+	nodesMap := make(map[string]*model.NodeInfo)
+	discoveredNodes, err := b.nodeDiscoverer.ListNodes(ctx)
```

Comment:
As I start looking at a persistent job store, would the job store be the correct place to implement something like `NodesForJob(jobID string) map[executionID]NodeID` to improve this lookup, or is there an issue with using job id in this context?

Reply:
That might make sense. I'll see what APIs we need after NodeManager's implementation